KIPs/kip-81.md

```diff
@@ -0,0 +1,773 @@
+
+- Each GC’s voter accounts \
+Stored in global mappings, and can be updated any time. Voter account mappings are updated by `refreshVoter()`
+
+#### Contract interface
+
+```solidity
+abstract contract StakingTracker {
+    struct Tracker {
+        // Tracker is only updated if trackStart <= block.number < trackEnd.
+        uint256 trackStart;
+        uint256 trackEnd;
+
+        // List of eligible nodes and their staking addresses.
+        // Initialized by createTracker().
+        address[] nodeIds;
+        mapping(address => address) rewardToNodeId;
+        mapping(address => address) stakingToNodeId;
+
+        // Balances and voting powers.
+        // Initialized by createTracker() and updated by refreshStake().
+        mapping(address => uint256) stakingBalances;
+        mapping(address => uint256) nodeBalances;
+        mapping(address => uint256) nodeVotes;
+        uint256 totalVotes;
+        uint256 eligibleNodes;
+    }
+
+    // Tracker objects. Added by createTracker().
+    mapping(uint256 => Tracker) private trackers; // trackerId => Tracker
+    uint256[] private allTrackerIds;
+
+    // 1-to-1 mapping between nodeId and voter account.
+    // Updated by refreshVoter().
+    mapping(address => address) private nodeIdToVoter; // nodeId => voterAddr
+    mapping(address => address) private voterToNodeId; // voterAddr => nodeId
+
+    /// @dev Creates a new Tracker and populate initial values from AddressBook
+    function createTracker(uint256 trackStart, uint256 trackEnd)
+        external returns(uint256 trackerId);
+
+    /// @dev Re-evaluate Tracker contents related to the staking contract
+    /// Anyone can call this function, but `staking` must be a staking contract
+    /// registered to the AddressBook.
+    function refreshStake(address staking) external;
+
+    /// @dev Re-evaluate voter account mapping related to the staking contract
+    /// Anyone can call this function, but `staking` must be a staking contract
+    /// registered to the AddressBook.
+    function refreshVoter(address staking) external;
+
+    /// @dev Return integer fields of a tracker
+    function getTrackerSummary(uint256 trackerId) external view returns(
+        uint256 trackStart,
+        uint256 trackEnd,
+        uint256 numNodes,
+        uint256 totalVotes,
+        uint256 eligibleNodes);
+
+    /// @dev Return balances and votes of all nodes stored in a tracker
+    function getAllTrackedNodes(uint256 trackerId) external view returns(
+        address[] memory nodeIds,
+        uint256[] memory nodeBalances,
+        uint256[] memory nodeVotes);
+
+    /// @dev Return the balance and votes of a specified node
+    function getTrackedNode(uint256 trackerId, address nodeId) external view returns(
+        uint256 nodeBalance,
+        uint256 nodeVotes);
+
+    /// @dev Find the node who owns the given voter account
+    function getNodeFromVoter(address voter) external view returns(address nodeId);
```

Comment:
What if two difference nodeIds have the same voter account? Do you have any logic to prevent this?

Reply:
Yes there is a logic to prevent this because node-voter is a 1-to-1 relation.
https://github.com/yeri-lee/kips/blob/33201dff2a121d0fa741edad113de3b0776bc9ea/KIPs/kip-81.md?plain=1#L378-L385
```solidity
        address newVoter = ICnStakingV2(staking).getVoterAddress();
        if (newVoter != address(0)) {
            require(voterToNodeId[newVoter] == address(0),
                    "Voter address already taken");
            voterToNodeId[newVoter] = nodeId;
            nodeIdToVoter[nodeId] = newVoter;
```

I've added a comment above `refreshVoter()` function. (https://github.com/klaytn/kips/commit/acfd52ebad8d79fd4e300bc034cf23dc501cba96)